### PR TITLE
os,osd: initial work to drop onreadable/onapplied callbacks

### PR DESCRIPTION
**Pull request by liewegas**
- adds tracking for in-flight writes, and wait_for_apply on read operations for FileStore
- removes lots of on_readable/applied callbacks in OSD
- more fixes to ceph_test_filestore_idempotent_sequence
- drops the "pinning" on handle_osd_map that is no longer necessary
- apply_transactions -> queue_transaction

The flush() operation is now only needed when there is an ordering needed between collections/sequencers. Split is the main example: the actual split happens on the parent collection, and we have to wait for that to flush before we start using the new child collection/sequencers.

There are still several on_applied users left:

- SnapMapper's OSDriver.  Here it keeps the in-memory cached mapping in place until it is applied and readable.  It's not necessary because reads would be correct (FileStore--and sometimes bluestore--would block), but it is faster.
- The OSD recovery code hasn't been changed to remove the applied callbacks yet.  Nothing really preventing us from doing that except that we need to be a bit careful because some of it might be used to throttle the recovery work.

...and probably a bit more.

Performance for FileStore is about the same.  Haven't check to see whether there is a measureable BlueStore boost yet.

### CODE REVIEW

File: src/os/ObjectStore.h [hunk 1]
```diff
@@ -117,101 +118,46 @@ class ObjectStore {
...
-   * Any transactions queued under a given sequencer will be applied in
-   * sequence.  Transactions queued under different sequencers may run
+   * Any transactions queued under a given collection will be applied in
+   * sequence.  Transactions queued under different collections may run
    * in parallel.
    *
-   * Clients of ObjectStore create and maintain their own Sequencer objects.
-   * When a list of transactions is queued the caller specifies a Sequencer to be used.
-   *
+   * ObjectStore users my get collection handles with open_collection() (or,
```
Comment (ifed01) [inline src/os/ObjectStore.h]: typo: my->may

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -6493,6 +6504,18 @@ ObjectStore::CollectionHandle BlueStore::open_collection(const coll_t& cid)
   return _get_collection(cid);
 }
 
+ObjectStore::CollectionHandle BlueStore::create_new_collection(
+  const coll_t& cid)
+{
+  RWLock::WLocker l(coll_lock);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Can be taken after collection creation.

**Comment by liewegas** [comment]
@rzarzynski i updated the last commit to fix a hang.. please retest!  thanks

**Comment by liewegas** [comment]
from @rzarzynski 
<pre>
I've got results for the latest commit. It's faster - the regression
in comparison to master is below 2% for randwrites: 1,93 and 1,39
respectively. :-)

Reference point [1] vs wip-kill-onreadable~ [2] vs the newest
wip-kill-onreadable [3]:

  seqwrite   4 KiB       "bw" :   30161 vs   30693 vs   30588,
  seqwrite  64 KiB       "bw" :  480848 vs  511888 vs  509298,
  randwrite  4 KiB       "bw" :  120254 vs  117502 vs  117914,
  randwrite 64 KiB       "bw" : 1510461 vs 1482712 vs 1489362.

[1] cb396a78d42f034dd61606a327cc703211e49cda
[2] 0f03c857c18475a97b80165f2696570730b208e5
[3] a7ab0faed16125527e7c3b5683e79da162d37006
</pre>

**Comment by liewegas** [comment]
My results are on my local box and pretty noisy.  Everything seems to speed up over time...  hrm..
<pre>
 base: 165252 157698 167683  (master)
foo-a: 170962 163602 162909  (wip-os-ch)
foo-b: 181918 174897 167302  (wip-os-ch + slow filestore tracking)
foo-c: 178579 164342 162355  (wip-os-ch + fast filestore tracking)
foo-d: 179306 176254 164431  (wip-os-ch + removal of some onreadable callbacks)
foo-e: 175176 179313 172316  (wip-os-ch + fast filestore tracking + removal of callbacks)
foo-f: 166346 174985 170836  (wip-os-ch + slow filestore tracking + removal of callbacks)
</pre>

**Comment by rzarzynski** [comment]
More results for a7ab0faed16125527e7c3b5683e79da162d37006 from incerta:
```
randwrite  4 KiB   117328,  116363,  116678,  117929,  115937
randwrite 64 KiB  1487512, 1482242, 1489939, 1487119, 1487347
seqwrite   4 KiB    30993,   31126,   30884,   31418,   30884
seqwrite  64 KiB   516277,  509546,  509979,  516726,  513052
```

**Comment by rzarzynski** [comment]
@liewegas: got the results for reads:

```
Reference point: cb396a78d42f034dd61606a327cc703211e49cda

randread,  4 KiB: 167427, 164852, 163403, 162503, 165675,
randread, 64 KiB: 1940050, 1944297, 1933320, 1938220, 1934929,
read,      4 KiB: 88122, 87789, 87821, 90193, 88627,
read,     64 KiB: 1099596, 1085303, 1095368, 1090855, 1089028,

wip-kill-onreadable: a7ab0faed16125527e7c3b5683e79da162d37006

randread,  4 KiB: 165292, 162285, 167443, 163996, 169634,
randread, 64 KiB: 1953484, 1928152, 1916863, 1933165, 1924817,
read,      4 KiB: 85601, 87416, 89353, 87323, 86968,
read,     64 KiB: 1084973, 1086560, 1092684, 1094833, 1087578,
```

**Comment by liewegas** [comment]
looks like 2-3% degradation on read. @jdurgin @gregsfortytwo @markhpc ok to proceed?

**Comment by gregsfortytwo** [comment]
That seems like a reasonable cost to pay for eliminating ~5k lines throughout the rest of the code base, though I didn't do a full review.

**Comment by liewegas** [comment]
@jdurgin this is passing tests now.  ready for review!

**Comment by liewegas** [comment]
@markhpc mind doing a final check on this for small reads and writes, filestore and bluestore?

**Comment by markhpc** [comment]
@liewegas on it

**Comment by markhpc** [comment]
NVMe tests are done though I also am running some (much slower to populate) HDD tests.  This is 72ad7640 vs wip-kill-onreadable 4k IO size to 4 NVMe backed OSDs.  High concurrency, 512GB total RBD volume size with 3x replication:


  | FS, 72ad7640 | FS, wip-kill-onreadable | BS, 72ad7640 | BS, wip-kill-onreadable
read | 728.1 | 735.3 | 332.5 | 338.3
write | 81.9 | 77.17 | 65.5518 | 77.9762
randread | 1225 | 1226 | 1093 | 1181
randwrite | 66.1807 | 64.0098 | 105.3 | 112.7
rw | 55.1074 | 56.7705 | 34.0918 | 39.3594
randrw | 63.0459 | 59.6758 | 102.8 | 108.9

In terms of percentage differences:

FS:

read: 0.99%
write: -6.13%
randread: 0.08%
randwrite: -3.39%
rw: 3.02%
randrw: -5.65%

BS:

read: 1.74%
write: 18.95%
randread: 8.05%
randwrite: 7.03%
rw: 15.45%
randrw: 5.93%

**Pull request merged**